src/keys/slip10.rs

```diff
@@ -399,88 +427,101 @@ impl<K: hazmat::Derivable> TryFrom<&[u8; 65]> for Slip10<K> {
     }
 }
 
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub struct Segment(pub u32);
-
-impl Segment {
-    pub fn is_hardened(&self) -> bool {
-        self.0 & Self::HARDEN_MASK != 0
+pub trait Segment: Copy + Into<u32> {
+    fn is_hardened(self) -> bool;
+    fn ser32(self) -> [u8; 4] {
+        self.into().to_be_bytes()
     }
-
-    pub fn bs(&self) -> [u8; 4] {
-        self.0.to_be_bytes() // ser32(i)
-    }
-
-    pub const HARDEN_MASK: u32 = 1 << 31;
+    fn harden(self) -> Hardened;
+    fn unharden(self) -> NonHardened;
 }
 
-impl From<u32> for Segment {
-    fn from(i: u32) -> Self {
-        Self(i)
-    }
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum SegmentHardeningError {
+    SegmentHardened,
+    SegmentNonHardened,
 }
 
-impl From<Segment> for u32 {
-    fn from(s: Segment) -> Self {
-        s.0
+impl From<SegmentHardeningError> for crate::Error {
+    fn from(inner: SegmentHardeningError) -> Self {
+        crate::Error::Slip10Error(inner)
     }
 }
 
-#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub struct Chain(Vec<Segment>);
+const HARDEN_MASK: u32 = 1 << 31;
 
-impl Chain {
-    pub fn empty() -> Self {
-        Self(Vec::new())
+impl Segment for u32 {
+    fn is_hardened(self) -> bool {
+        self & HARDEN_MASK != 0
     }
-
-    pub fn from_segments<'a, I: IntoIterator<Item = &'a Segment>>(is: I) -> Self {
-        Self(is.into_iter().cloned().collect())
+    fn harden(self) -> Hardened {
+        Hardened(self | HARDEN_MASK)
     }
-
-    pub fn from_u32<I: IntoIterator<Item = u32>>(is: I) -> Self {
-        Self(is.into_iter().map(Segment).collect())
+    fn unharden(self) -> NonHardened {
+        NonHardened(self & !HARDEN_MASK)
     }
+}
 
-    pub fn from_u32_hardened<I: IntoIterator<Item = u32>>(is: I) -> Self {
-        Self::from_u32(is.into_iter().map(|i| Segment::HARDEN_MASK | i))
-    }
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
+pub struct Hardened(pub u32);
```

Comment:
hmm ... do we need the inner to be `pub` when we have a fallible `TryFrom`. This could be misused and create an invalid `Hardened`, right?

Reply:
Right. Copy-paste error, didn't notice myself. Fixed, inner values are private.

Reply:
Good point. This should probably be `pub(crate)`